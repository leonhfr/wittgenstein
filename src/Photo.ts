// Internal.
import * as Types from './types';

// Code.
export class Photo {
  readonly id: string;
  readonly owner: string;
  readonly secret: string;
  readonly server: string;
  readonly farm: number;
  readonly title: string;
  readonly description: string;
  readonly ownername: string;
  readonly views: number;
  readonly tags: Array<string>;
  readonly latitude: number;
  readonly longitude: number;
  readonly context: number;
  readonly zoneId: string;
  readonly inside: boolean;

  static create(input: unknown): Photo | Error {
    if (isPhoto(input)) {
      return new Photo(input);
    }
    const errMsg = isSafePhoto(input).errMsg;
    return new Error(`ValidationError: ${errMsg}`);
  }

  private constructor(input: CreatePhotoInput) {
    this.id = input.id;
    this.owner = input.owner;
    this.secret = input.secret;
    this.server = input.server;
    this.farm = input.farm;
    this.title = input.title;
    this.description = input.description;
    this.ownername = input.ownername;
    this.views = input.views;
    this.tags = input.tags;
    this.latitude = input.latitude;
    this.longitude = input.longitude;
    this.context = input.context;
    this.zoneId = input.zoneId;
    this.inside = input.inside;
  }
}

// Validation.

export const PHOTO_PROPS = [
  'id',
  'owner',
  'secret',
  'server',
  'farm',
  'title',
  'description',
  'ownername',
  'views',
  'tags',
  'latitude',
  'longitude',
  'context',
  'zoneId',
  'inside',
];

export const isPhoto = (input: unknown): input is CreatePhotoInput => {
  return isSafePhoto(input).isSafe;
};

export const isSafePhoto = (input: unknown): Types.IsSafe => {
  if (typeof input !== 'object') {
    return {
      isSafe: false,
      errMsg: `Expected type of input to be object, got ${typeof input}.`,
    };
  }

  if (input === null) {
    return {
      isSafe: false,
      errMsg: `Expected input not to be null.`,
    };
  }

  for (const key of PHOTO_PROPS) {
    if (!input.hasOwnProperty(key)) {
      return {
        isSafe: false,
        errMsg: `Expected all input properties to be set, missing ${key}.`,
      };
    }
  }

  const {
    id,
    owner,
    secret,
    server,
    farm,
    title,
    description,
    ownername,
    views,
    tags,
    latitude,
    longitude,
    context,
    zoneId,
    inside,
  } = input as {
    id: unknown;
    owner: unknown;
    secret: unknown;
    server: unknown;
    farm: unknown;
    title: unknown;
    description: unknown;
    ownername: unknown;
    views: unknown;
    tags: unknown;
    latitude: unknown;
    longitude: unknown;
    context: unknown;
    zoneId: unknown;
    inside: unknown;
  };

  const booleans = { inside };

  const numbers = { farm, views, latitude, longitude, context };

  const strings = {
    id,
    owner,
    secret,
    server,
    title,
    description,
    ownername,
    zoneId,
  };

  const stringArrays = { tags };

  // Booleans

  for (const [key, value] of Object.entries(booleans)) {
    if (typeof value !== 'boolean') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be boolean, got ${typeof value}.`,
      };
    }
  }

  // Numbers

  for (const [key, value] of Object.entries(numbers)) {
    if (typeof value !== 'number') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be number, got ${typeof value}.`,
      };
    }
  }

  // Strings

  for (const [key, value] of Object.entries(strings)) {
    if (typeof value !== 'string') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be string, got ${typeof value}.`,
      };
    }
  }

  // Array<string>

  for (const [key, value] of Object.entries(stringArrays)) {
    if (!Array.isArray(value)) {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be array, got ${typeof value}.`,
      };
    }

    for (const item of value) {
      if (typeof item !== 'string') {
        return {
          isSafe: false,
          errMsg: `Expected type of input.${key} elements to be string, got ${typeof item} for one of them.`,
        };
      }
    }
  }

  return {
    isSafe: true,
    errMsg: '',
  };
};

// Interface

export interface CreatePhotoInput {
  readonly id: string;
  readonly owner: string;
  readonly secret: string;
  readonly server: string;
  readonly farm: number;
  readonly title: string;
  readonly description: string;
  readonly ownername: string;
  readonly views: number;
  readonly tags: Array<string>;
  readonly latitude: number;
  readonly longitude: number;
  readonly context: number;

  readonly zoneId: string;
  readonly inside: boolean;
}
