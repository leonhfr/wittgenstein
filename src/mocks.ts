// Internal.
import { CreateAreaInput } from './Area';
import { CreatePhotoInput } from './Photo';

// Code.
export const createMockAreaInput = (area?: any): CreateAreaInput => {
  const _area = area || {};
  return {
    id: _area.id || '8ea8f412-4ee8-4f1d-9228-4c133b8fbd37',
    name: _area.name || 'Sant Marti',
    file: _area.file || 'sant-marti',
    lastScheduledAt: _area.lastScheduledAt || 0,
    refreshRate: _area.refreshRate || 86400,
    enabled: _area.enabled || true,
    zonesComputed: _area.zonesComputed || false,
  };
};

export const createMockPhotoInput = (photo?: any): CreatePhotoInput => {
  const _photo = photo || {};
  return {
    id: _photo.id || '32457360587',
    owner: _photo.owner || '41710899@N08',
    secret: _photo.secret || '69b27ceb18',
    server: _photo.server || '4898',
    farm: _photo.farm || 5,
    title: _photo.title || 'Arc de Triomf',
    description:
      _photo.description ||
      'The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.',
    ownername: _photo.ownername || 'jdf_92',
    views: _photo.views || 16,
    tags: _photo.tags || [
      'spain',
      'barcelona',
      'arcdetriomf',
      '1888barcelonaworldfair',
      'neomudéjar',
      '1888barcelonauniversalexposition',
      'catelonia',
    ],
    latitude: _photo.latitude || 41.390791,
    longitude: _photo.longitude || 2.18114,
    context: _photo.context || 0,
  };
};
