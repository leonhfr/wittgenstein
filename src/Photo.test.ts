// Internal.
import { Photo, PHOTO_PROPS, isPhoto } from './Photo';
import * as Mocks from './mocks';

// Code.
describe('Photo', () => {
  describe('Class', () => {
    it('should be defined', () => {
      expect(Photo).toBeDefined();
    });

    describe('create', () => {
      it('should fail to create a new instance - invalid input', () => {
        expect(Photo.create(null)).toMatchSnapshot();
      });
      it('should create a new instance', () => {
        expect(Photo.create(Mocks.createMockPhotoInput())).toMatchSnapshot();
      });
    });
  });

  describe('Validation', () => {
    it('should be defined', () => {
      expect(PHOTO_PROPS).toBeDefined();
    });
    it('should match the expected value', () => {
      expect(PHOTO_PROPS).toMatchSnapshot();
    });
    it('should be up to date', () => {
      expect(PHOTO_PROPS).toEqual(
        Object.keys(Photo.create(Mocks.createMockPhotoInput()))
      );
    });
  });

  describe('isPhoto', () => {
    it('should be defined', () => {
      expect(isPhoto).toBeDefined();
    });
    it('should handle an invalid input - wrong type', () => {
      expect(isPhoto('__FAIL__')).toBe(false);
    });
    it('should handle an invalid input - null type', () => {
      expect(isPhoto(null)).toBe(false);
    });
    it('should handle an invalid input - missing props', () => {
      expect(isPhoto({})).toBe(false);
    });
    it('should handle an invalid input - numbers', () => {
      expect(
        isPhoto({
          ...Mocks.createMockPhotoInput(),
          views: '__FAIL__',
        })
      ).toBe(false);
    });
    it('should handle an invalid input - strings', () => {
      expect(
        isPhoto({
          ...Mocks.createMockPhotoInput(),
          id: false,
        })
      ).toBe(false);
    });
    it('should handle an invalid input - arrays', () => {
      expect(
        isPhoto({
          ...Mocks.createMockPhotoInput(),
          tags: '__FAIL__',
        })
      ).toBe(false);
    });
    it('should handle an invalid input - string arrays', () => {
      expect(
        isPhoto({
          ...Mocks.createMockPhotoInput(),
          tags: [false],
        })
      ).toBe(false);
    });
  });
});
