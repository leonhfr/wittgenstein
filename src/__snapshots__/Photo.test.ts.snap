// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo Class create should create a new instance 1`] = `
Photo {
  "context": 0,
  "description": "The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.",
  "farm": 5,
  "id": "32457360587",
  "inside": true,
  "latitude": 41.390791,
  "longitude": 2.18114,
  "owner": "41710899@N08",
  "ownername": "jdf_92",
  "secret": "69b27ceb18",
  "server": "4898",
  "tags": Array [
    "spain",
    "barcelona",
    "arcdetriomf",
    "1888barcelonaworldfair",
    "neomudéjar",
    "1888barcelonauniversalexposition",
    "catelonia",
  ],
  "title": "Arc de Triomf",
  "views": 16,
  "zoneId": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
}
`;

exports[`Photo Class create should fail to create a new instance - invalid input 1`] = `[Error: ValidationError: Expected input not to be null.]`;

exports[`Photo Validation should match the expected value 1`] = `
Array [
  "id",
  "owner",
  "secret",
  "server",
  "farm",
  "title",
  "description",
  "ownername",
  "views",
  "tags",
  "latitude",
  "longitude",
  "context",
  "zoneId",
  "inside",
]
`;
